README
//***************************//
//*  Christopher J. Prokop  *//
//*  cprokop@lanl.gov       *//
//*  Cathleen E. Fry        *//
//*  cfry@lanl.gov          *//
//*  README                 *// 
//*  Last Edit: 03/03/20   *//  
//***************************//

The purpose of this document is to explain how the program works.


1) Prerequisites 

   At the same level as the DANCE_Analysis directory make two directories for Time Deviations and Energy Calibrations.  The names of these don't matter but they have to match what is being looked for by the code in global.h

   In the Calibrations directory sym link the calib_ideal.dat file

   One will need to make symbolic links for:
   'stage0_root' (the location of the root output from stage 0), 
   'stage0_bin' (the location of the binary output from stage 0 and potentially the input for stage 1), 
   'stage1_root' (the location of the root output from stage 1), and 
   'stage1_bin' (if wanted the binary output from stage 1)
   'stage0_simulated' (the location of .bin files made by the GEANT4 simulation)

   There also needs to be a symbolic link or local directory containing MIDAS data

   Compile the code by typing make.  


2) Running the DANCE Analysis 

   The code is invoked by the following command and arguments:

FOR STAGE0 and STAGE1 ANALYSIS

   "./DANCE_Analysis /pathtodata runnumber (optional subrunnumber) cfgfile.cfg"
   
   The argument 'pathtodata' is either a symbolic link to the location or the location of the .mid or .mid.gz files. 

   The runnumber is padded with 6 zeros and other padding formats would require a minor change in main.cpp

   The subrunnumber is padded with 3 zeros and other padding formats would require a minor change in main.cpp, if a subrunnumber is not specified, all subruns of a run are sorted

   Sample .cfg files are included when you clone the package and can be modified thereafter.


FOR SIMULATION ANLYSIS

   "./DANCE_Analysis cfgfile.cfg"

   Here the path to the data is whatever the symbolic link "stage0_simulated" points to


3) What Stage 0 does

   Stage 0 is responsible for unpacking and time sorting the data from either MIDAS format and performing analysis on the waveforms.  A variety of diagnostic histograms are made, time deviations are calculated, and a output binary file of the salient quantities of the data is produced.  These quantities are Ifast (the fast or short integral from the PSD firmware), Islow (the slow or long integral from the PSD firmware), the Time-Of-Flight (this is a bit of a misnomer because its actually the 31 bits of the trigger timestamp + the 16 bits (shifted up 31 bits) of the extended time stamp + the CFD timestamp from the waveform analysis), and the ID obtained from mapping the board and channel with the DANCE map.


4) What Stage 1 does

   Stage 1 is responsible for physics analysis.  The binary format written by stage 0 is read in and events are built for DANCE, T0, and the beam monitors.  More to come on this later.  For pileup corrections right now, do stage 1 from midas, not binary.  Updates to come on this. To do pileup corrections, run with #define Histogram_Waveforms, then edit the gate in global.h. Be careful with your pileup gate, and know that not all histograms that are gated on no pileup have explicit labeling. 



Release_Notes
//***************************//
//*  Christopher J. Prokop  *//
//*  cprokop@lanl.gov       *//
//*  Cathleen E. Fry        *//
//*  cfry@lanl.gov          *//
//*  Release_Notes          *// 
//*  Last Edit: 01/24/20    *//  
//***************************//

The purpose of this document is to allow users to more efficiently track and understand changes made to the DANCE Analysis software.


Version 3.0.0 - Start of the release notes.  Added the caen2018 data format to the unpacker.  Requires a statement of 'DataFormat caen2018' or 'DataFormat caen2015' to be placed in the .cfg file.  

Version 3.0.1 - Minor efficieny upgrades to waveform unpacking in the caen2018 data format.

Version 3.1.0 - Changed the way the Crystal and DEvent Blocking Times are done.  Before this change the last_timestamp variable in each respective case was updated regardless of the event being valid or invalid.  This is incorrect because a blocked entry or event should not be recorded in the first place and thus the timestamp shouldn't update.

Version 3.1.1 - Added unpacking of 0x8500 + 4n diagnostic register in the scalers

Version 3.1.2 - Added checking of bit 26 (CAEN Board Fail Bit) in the second board aggregate header word.  Result is a warning if it is found high.  Also removed unpacking of channelmask2 since it does not make sense with the data format of our hardware.

Version 3.1.3 - Added time between beam monitor events to look into possible dead time for beam monitors

Version 3.1.4 - Fixed #define statements in the scaler part of the caen2018 unpacker.  They were outside of brackets of for loops causing the code to not compile when #define ScalerVerbose was off

Version 3.2.0 - Added output txt file for diagnostics information in caen2018 format.

Version 3.3.0 - Changed DANCE, U235, and Li6 flight path lengths and time offsets.

Version 3.4.0 - Added waveform hisograms for all IDs in one TH3I. X_Y_Z SampleNumber_WaveformValue_ID.  Also added the hID_Raw spectrum which fills the number of counts for each channel +board*16 up to 256 (14 boards).  This aids in debugging the channel drop problem for non dance crystal channels.  Slight changes to the order of output root file creation in main.cpp

Version 3.4.1 - Changed the location of the Q-Value gates to the .cfg file for stage 1. Hardcoded maximum of 10 Q-value gates.  Can't imagine a reasonable scenario where someone needs more.  To change this the number of TH3F histograms needs to be changed and the array size of QGates in main.cpp needs to be changed. 

Version 3.4.2 - Added the path for the TimeDeviations and Calibrations to global.h and moved the folders one directory back for my version of the code.

Version 3.5.0 - Changed the unpacker to reflect the new caen2018 data format with the "user extras" word and the placeholders for the vme and hv event types. 

Version 4.0.0 - Added the ability to run simulated data through the analysis.  More features added to hCrID_Diagnostics in utilities.

Version 4.1.0 - Added the vme scalers to the unpacker for both caen2015 and caen2018 formats. Scaler description file is now in the Config directory.

Version 4.1.1 - Changed the verobosity flags for scalers adding on for diagnostics. 

Version 5.0.0 - Changed the way the main is invoked removing extra command line options.  For Stage0: \"./DANCE_Analysis pathtodata runnumber cfgfile.cfg \",for Stage1: \"./DANCE_Analysis runnumber cfgfile.cfg \" and for Simulations: \"./DANCE_Analysis cfgfile.cfg \".  Minor changes in unpacker so variables that dont matter at a particular stage are not output to screen.

Version 5.1.0 - Changed the time deviations such that it looks for a wider window for the mean and progressively closes in.  It was possible for low stats channels with larger backgrounds that it missed entirely.   


main.cpp

////////////////////////////////////////////////////////////////////////
//                                                                    //
//   Software Name: DANCE Data Acquisition and Analysis Package       //
//     Subpackage: DANCE_Analysis                                     //
//   Identifying Number: C18105                                       // 
//                                                                    //
////////////////////////////////////////////////////////////////////////
//                                                                    //
//                                                                    //
// Copyright 2019.                                                    //
// Triad National Security, LLC. All rights reserved.                 //
//                                                                    //
//                                                                    //
//                                                                    //
// This program was produced under U.S. Government contract           //
// 89233218CNA000001 for Los Alamos National Laboratory               //
// (LANL), which is operated by Triad National Security, LLC          //
// for the U.S. Department of Energy/National Nuclear Security        //
// Administration. All rights in the program are reserved by          //
// Triad National Security, LLC, and the U.S. Department of           //
// Energy/National Nuclear Security Administration. The Government    //
// is granted for itself and others acting on its behalf a            //
// nonexclusive, paid-up, irrevocable worldwide license in this       //
// material to reproduce, prepare derivative works, distribute        //
// copies to the public, perform publicly and display publicly,       //
// and to permit others to do so.                                     //
//                                                                    //
// This is open source software; you can redistribute it and/or       //
// modify it under the terms of the GPLv2 License. If software        //
// is modified to produce derivative works, such modified             //
// software should be clearly marked, so as not to confuse it         //
// with the version available from LANL. Full text of the GPLv2       //
// License can be found in the License file of the repository         //
// (GPLv2.0_License.txt).                                             //
//                                                                    //
////////////////////////////////////////////////////////////////////////

//***************************//
//*  Christopher J. Prokop  *//
//*  cprokop@lanl.gov       *//
//*  Cathleen E. Fry        *//
//*  cfry@lanl.gov          *//
//*  main.cpp               *// 
//*  Last Edit: 01/24/20    *//  
//***************************//

//File includes
#include "global.h"
#include "main.h"
#include "unpacker.h"
#include "analyzer.h"
#include "calibrator.h"
#include "validator.h"
#include "eventbuilder.h"

//Root include
#include "TROOT.h"

//C/C++ includes
#include <sys/time.h>
#include <queue>

using namespace std;

stringstream mmsg;

int main(int argc, char *argv[]) {

  int func_ret=0;

  const char* rootver=gROOT->GetVersion();
  
  if (rootver[0]!='6'){
    if(rootver[0]=='5' && rootver[2]!='3'){
       DANCE_Error("Main","Wrong root version, go to 5.34/36 or 6, otherwise time deviations won't make correctly");
       return -1;
    }
  }

  //Read in the version information
  ifstream version_file;
  version_file.open(".version");
  
  string filler, version;
  version_file >> filler >> version;
  

  //Output the version information
  mmsg.str("");
  mmsg<<"DANCE Analyzer Version "<<version<<" Initializing";
  DANCE_Init("Main",mmsg.str());

  Analysis_Parameters analysis_params;
  for(int eye=0; eye<256; eye++) {
    analysis_params.last_timestamp[eye]=0;
    analysis_params.last_Islow[eye]=65535;
    analysis_params.last_Eslow[eye]=20;
    analysis_params.last_Efast[eye]=20;

    analysis_params.last_valid_timestamp[eye]=0;
    analysis_params.last_valid_Islow[eye]=65535;
    analysis_params.last_valid_Eslow[eye]=20;

    analysis_params.entries_unpacked=0;
    analysis_params.entries_awaiting_timesort=0;
    analysis_params.entries_written_to_binary=0;
    analysis_params.entries_processed=0;
    analysis_params.entries_invalid=0;
    analysis_params.entries_built=0;
    analysis_params.events_built=0;

    analysis_params.entries_analyzed=0;
    analysis_params.DANCE_entries_analyzed=0;
    analysis_params.T0_entries_analyzed=0;
    analysis_params.He3_entries_analyzed=0;
    analysis_params.Li6_entries_analyzed=0;
    analysis_params.Bkg_entries_analyzed=0;
    analysis_params.U235_entries_analyzed=0;
    analysis_params.Unknown_entries=0;

    analysis_params.events_analyzed=0;
    analysis_params.DANCE_events_analyzed=0;
    analysis_params.T0_events_analyzed=0;
    analysis_params.He3_events_analyzed=0;
    analysis_params.Li6_events_analyzed=0;
    analysis_params.Bkg_events_analyzed=0;
    analysis_params.U235_events_analyzed=0;

    analysis_params.max_buffer_utilization=0;
    
    analysis_params.first_sort=true;
    analysis_params.event_building_active=false; 
    analysis_params.smallest_timestamp=2.814749767e14; //this is the largest the clock can be on CAEN boards
    analysis_params.largest_timestamp=0;
    analysis_params.largest_subrun_timestamp=0;
    analysis_params.last_subrun_timestamp=0;
  }
  
  Input_Parameters input_params;

  //Initialize input parameters
  input_params.Crystal_Blocking_Time=0;
  input_params.DEvent_Blocking_Time=0;
  input_params.Read_Binary=false;
  input_params.Write_Binary=false;
  input_params.Read_Simulation=false;
  input_params.SingleSubrun=false;
  input_params.Coincidence_Window=10;
  input_params.HAVE_Threshold=false;
  input_params.Energy_Threshold=0.15; //MeV
  input_params.FitTimeDev=0;
  input_params.QGatedSpectra=false;
  input_params.NQGates=0;
  input_params.IsomerSpectra=false;
  input_params.NIsomers=0;
  input_params.Evaluate_DeadTime=false;
  input_params.Artificial_TOF=0;
  input_params.Long_Gate=1000;
  input_params.Use_Firmware_FineTime=false;
  input_params.Analysis_Stage = 0;
  input_params.Buffer_Depth = 10;
      
  //Control things
  int RunNum=0;
  int SubRunNum=0;
  string pathtodata;
  string cfgfile;
  
  //Make sure the number of arguments is reasonable
  if(argc==2) {
    cfgfile = argv[1];
  }
  else if(argc==4) {
    pathtodata = argv[1];
    RunNum = atoi(argv[2]);
    cfgfile = argv[3];
  }
  else if(argc==5) {
    pathtodata = argv[1];
    RunNum = atoi(argv[2]);
    cfgfile = argv[4];
    SubRunNum = atoi(argv[3]);
    input_params.SingleSubrun = true;
  }

  else {
    DANCE_Error("Main","Too many or too few arguments provided.  See README file");
    DANCE_Error("Main","for Stage 0 and Stage 1: \"./DANCE_Analysis pathtodata runnumber (optional subrunnumber) cfgfile.cfg");
    DANCE_Error("Main","for Simulations: \"./DANCE_Analysis cfgfile.cfg");
    return -1;
  }

  //Set the run number
  input_params.RunNumber = RunNum;
  input_params.NumSubRun = 0;

  //Display the run number
  mmsg.str("");
  mmsg<<"Run Number: "<<input_params.RunNumber;
  DANCE_Info("Main",mmsg.str());

  DANCE_Info("Main","Opening Configuration File");
  
  ifstream cfgf;
  cfgf.open(cfgfile.c_str());

  if(cfgf.is_open()) {
    
    mmsg.str("");
    mmsg<<"Configuration File "<<cfgfile.c_str()<<" is Open";
    DANCE_Success("Main",mmsg.str());

    string item;
    while(!cfgf.eof()) {
      cfgf>>item;
      if(item.compare("Coincidence_Window") == 0) {
      	cfgf>>input_params.Coincidence_Window;
      }
      if(item.compare("Read_Binary") == 0) {
      	cfgf>>input_params.Read_Binary;
      }
      if(item.compare("Write_Binary") == 0) {
	cfgf>>input_params.Write_Binary;
      }      
      if(item.compare("Read_Simulation") == 0) {
	cfgf>>input_params.Read_Simulation;
      }      
      if(item.compare("Simulation_File_Name") == 0) {
	cfgf>>input_params.Simulation_File_Name;
      } 
      if(item.compare("Crystal_Blocking_Time") == 0) {
	cfgf>>input_params.Crystal_Blocking_Time;
      }    
      if(item.compare("DEvent_Blocking_Time") == 0) {
	cfgf>>input_params.DEvent_Blocking_Time;
      } 
      if(item.compare("HAVE_Threshold") == 0) {
	cfgf>>input_params.HAVE_Threshold;
      } 
      if(item.compare("Energy_Threshold") == 0) {
	cfgf>>input_params.Energy_Threshold;
      }  
      if(item.compare("FitTimeDev") == 0) {
	cfgf>>input_params.FitTimeDev;
      } 
      if(item.compare("DataFormat") == 0) {
	cfgf>>input_params.DataFormat;
      }
      if(item.compare("NQGates") == 0) {
	cfgf>>input_params.NQGates;
	for(int eye=0; eye<2*input_params.NQGates; eye++) {
	  cfgf >> input_params.QGates[eye];
	}
      }
      if(item.compare("NIsomers") == 0) {
	cfgf>>input_params.NIsomers;
	for(int eye=0; eye<input_params.NIsomers; eye++) {
	  cfgf >> input_params.IsomerPromptQGates[2*eye] >> input_params.IsomerPromptQGates[2*eye+1];
	  cfgf >> input_params.IsomerPromptMclGates[2*eye] >> input_params.IsomerPromptMclGates[2*eye+1];
	  cfgf >> input_params.IsomerPromptTOFGates[2*eye] >> input_params.IsomerPromptTOFGates[2*eye+1];
	  cfgf >> input_params.IsomerDelayedQGates[2*eye] >> input_params.IsomerDelayedQGates[2*eye+1];
	  cfgf >> input_params.IsomerDelayedMclGates[2*eye] >> input_params.IsomerDelayedMclGates[2*eye+1];
	  cfgf >> input_params.IsomerDelayedTOFGates[2*eye] >> input_params.IsomerDelayedTOFGates[2*eye+1];
	}
      }

      if(item.compare("JMOD_Background") == 0) {
	cfgf>>input_params.JMOD_Background;
      }

      if(item.compare("Evaluate_DeadTime") == 0) {
	cfgf>>input_params.Evaluate_DeadTime;
      }   

      if(item.compare("Artificial_TOF") == 0) {
	cfgf>>input_params.Artificial_TOF;
      }   
      if(item.compare("DetectorLoad_FileName") == 0) {
	cfgf>>input_params.DetectorLoad_FileName;
      }   
      if(item.compare("DetectorLoad_HistName") == 0) {
	cfgf>>input_params.DetectorLoad_HistName;
      }   
      if(item.compare("Long_Gate") == 0) {
	cfgf>>input_params.Long_Gate;
      }  
      if(item.compare("Use_Firmware_FineTime") == 0) {
	cfgf>>input_params.Use_Firmware_FineTime;
      }  
      if(item.compare("Analysis_Stage") == 0) {
	cfgf>>input_params.Analysis_Stage;
      }
      if(item.compare("Buffer_Depth") == 0) {
	cfgf>>input_params.Buffer_Depth;
      } 
   
    }

    //Set the bool for QGates
    if(input_params.NQGates>0) {
      input_params.QGatedSpectra = true;
    }
    else {
      input_params.QGatedSpectra = false;
    }
    
    //Set the bool for Isomers
    if(input_params.NIsomers>0) {
      input_params.IsomerSpectra = true;
    }
    else {
      input_params.IsomerSpectra = false;
    }
    
    mmsg.str("");
    mmsg<<"Read Configuration File: "<<cfgfile;
    DANCE_Success("Main",mmsg.str());

    cout<<"Analysis Stage: "<<input_params.Analysis_Stage<<endl;
    cout<<"Coincidence Window: "<<input_params.Coincidence_Window<<endl;
    cout<<"Read Binary: "<<input_params.Read_Binary<<endl;
    cout<<"Write Binary: "<<input_params.Write_Binary<<endl;
    cout<<"Read Simulation: "<<input_params.Read_Simulation<<endl;
    if(input_params.Read_Simulation) {
      cout<<"Simulated File Name: "<<input_params.Simulation_File_Name<<endl;
    }
 
    cout<<"Buffer Depth: "<<input_params.Buffer_Depth<<" seconds"<<endl;
     
    cout<<"Crystal Blocking Time: "<<input_params.Crystal_Blocking_Time<<endl;
    cout<<"DANCE Event Blocking Time: "<<input_params.DEvent_Blocking_Time<<endl;
    cout<<"Have Threshold: "<<input_params.HAVE_Threshold<<endl;
    cout<<"Energy Threshold: "<<input_params.Energy_Threshold<<endl;
    cout<<"Fit Time Deviations: "<<input_params.FitTimeDev<<endl;
    cout<<"Data Format: "<<input_params.DataFormat<<endl;
    cout<<"Number of Q-Value Gates: "<<input_params.NQGates<<endl;

    for(int eye=0; eye<input_params.NQGates; eye++) {
      cout<<"Q-Value Gate "<<eye<<": "<<input_params.QGates[2*eye]<<" to "<<input_params.QGates[2*eye+1]<<" MeV"<<endl;
    }

    for(int eye=0; eye<input_params.NIsomers; eye++) {
      cout<<"Isomer Gate "<<eye<<endl;
      cout<<"Prompt :";
      cout<<"  ESum from: "<<input_params.IsomerPromptQGates[2*eye]<<" to "<<input_params.IsomerPromptQGates[2*eye+1]<<" MeV";
      cout<<"  Mcl from: "<<input_params.IsomerPromptMclGates[2*eye]<<" to "<<input_params.IsomerPromptMclGates[2*eye+1];
      cout<<"  TOF from: "<<input_params.IsomerPromptTOFGates[2*eye]<<" to "<<input_params.IsomerPromptTOFGates[2*eye+1]<<" ns"<<endl;
      cout<<"Delayed :";
      cout<<"  ESum from: "<<input_params.IsomerDelayedQGates[2*eye]<<" to "<<input_params.IsomerDelayedQGates[2*eye+1]<<" MeV";
      cout<<"  Mcl from: "<<input_params.IsomerDelayedMclGates[2*eye]<<" to "<<input_params.IsomerDelayedMclGates[2*eye+1];
      cout<<"  TOF from: "<<input_params.IsomerDelayedTOFGates[2*eye]<<" to "<<input_params.IsomerDelayedTOFGates[2*eye+1]<<" ns"<<endl;	 
    }
    cout<<"Long Gate (\"Minimum Time Between Crystal Hits\"): "<< input_params.Long_Gate<<endl;
    cout<<"Use Firmware Fine Time: "<<input_params.Use_Firmware_FineTime<<endl;
  }
  
  //If no configuration file then exit
  else {
    mmsg.str("");
    mmsg<<"Failed to Read Configuration File: "<<cfgfile;
    DANCE_Error("Main",mmsg.str());
    return -1;
  }
 
  //make the file handle
  gzFile gz_in;
  queue<gzFile> gz_queue;

  
  //Figure out what we are reading in.
  //MIDAS Files have a .mid or .mid.gz ending and staged analysis files are .bin or .bin.gz

  //Name of the midas or bin file
  stringstream runname;
  runname.str();
    
  //MIDAS input
  if(input_params.Read_Binary == 0 && input_params.Read_Simulation == 0) {

    stringstream midasrunname;
    midasrunname.str();
    stringstream midassubrunname;
    midassubrunname.str();

    midasrunname << pathtodata << "/run" << std::setfill('0') << std::setw(6) << RunNum << ".mid";
    midassubrunname << pathtodata << "/run" << std::setfill('0') << std::setw(6) << RunNum << "_";
    if (input_params.SingleSubrun){
      midassubrunname << std::setw(3) << SubRunNum;
    }
    else {
      midassubrunname << std::setw(3) << 0;
    } 
    midassubrunname << ".mid";

    mmsg.str("");
    mmsg<<"Checking for: "<<midassubrunname.str();
    DANCE_Info("Main",mmsg.str());
    
    //Look for uncompressed .mid subrun files
    gz_in=gzopen(midassubrunname.str().c_str(),"rb");
    
    //check to see if its open
    if (input_params.SingleSubrun){
       if(gz_in) {
         mmsg.str("");
         mmsg<<"File "<<midassubrunname.str().c_str()<<" Found";
         DANCE_Success("Main",mmsg.str());
          
         runname << midassubrunname.str();
         gz_queue.push(gz_in);
      }
      else { //particular subrun gz
        midassubrunname << ".gz";         
        mmsg<<"Checking for: "<<midassubrunname.str();
        DANCE_Info("Main",mmsg.str());
        input_params.SubRunNumber=0;
        
        gz_in=gzopen(midassubrunname.str().c_str(),"rb");
        if (gz_in) {
          mmsg.str("");
          mmsg<<"File "<<midassubrunname.str().c_str()<<" Found";
          DANCE_Success("Main",mmsg.str());
          
          runname << midassubrunname.str();
          gz_queue.push(gz_in);
        }
      }
    }//end if single subrun
    else {//all the subruns in a run 
        if(gz_in) {//subruns, not zipped
          mmsg.str("");
          mmsg<<"File "<<midassubrunname.str().c_str()<<" Found";
          DANCE_Success("Main",mmsg.str());
          
          runname << midassubrunname.str();
          input_params.SubRunNumber=0;
 
          while(gz_in) {
            input_params.NumSubRun++;
            gz_queue.push(gz_in);
            midassubrunname.str("");
            midassubrunname << pathtodata << "/run" << std::setfill('0') << std::setw(6) << RunNum << "_" << std::setw(3) << input_params.NumSubRun << ".mid";
            gz_in=gzopen(midassubrunname.str().c_str(),"rb");
          }
        }
      else {//subruns zipped
        //look for compressed .mid.gz subrun files
        midassubrunname << ".gz";
        
        mmsg.str("");
        mmsg<<"Checking for: "<<midassubrunname.str();
        DANCE_Info("Main",mmsg.str());
        input_params.SubRunNumber=0;
        
        gz_in=gzopen(midassubrunname.str().c_str(),"rb");
        if(gz_in) {
          mmsg.str("");
          mmsg<<"File "<<midassubrunname.str().c_str()<<" Found";
          DANCE_Success("Main",mmsg.str());

          runname << midassubrunname.str();
          input_params.SubRunNumber=0;
          while(gz_in) {
            input_params.NumSubRun++;
            gz_queue.push(gz_in);
            midassubrunname.str("");
            midassubrunname << pathtodata << "/run" << std::setfill('0') << std::setw(6) << RunNum << "_" << std::setw(3) << input_params.NumSubRun << ".mid.gz";
            gz_in=gzopen(midassubrunname.str().c_str(),"rb");
          }
        }
        else { //look for uncompressed .mid files (no subrun)
          gz_in=gzopen(midasrunname.str().c_str(),"rb");
        
          if(gz_in) {
            mmsg.str("");
            mmsg<<"File "<<midasrunname.str().c_str()<<" Found";
            DANCE_Success("Main",mmsg.str());
            runname << midasrunname.str();
            input_params.SubRunNumber=-1;
            gz_queue.push(gz_in);
          }
          else { //look for .mid.gz files (no subrun)
            midasrunname << ".gz";
            gz_in=gzopen(midasrunname.str().c_str(),"rb");
          
            if(gz_in) {
              mmsg.str("");
              mmsg<<"File "<<midasrunname.str().c_str()<<" Found";
              DANCE_Success("Main",mmsg.str());
              runname << midasrunname.str();
              input_params.SubRunNumber=-1;
              gz_queue.push(gz_in);
            }
          }
        }
      } //end .mid.gz
    } //end not single subrun
  } //end read midas 

  //Binary input that is not simulation
  else if(input_params.Read_Binary == 1 && input_params.Read_Simulation == 0) {
    
    stringstream binaryrunname;
    binaryrunname.str();
    binaryrunname << pathtodata << "/stage0_run_" << RunNum << ".bin";
    stringstream binarysubrunname;
    binarysubrunname.str();
    binarysubrunname << pathtodata << "/stage0_run_" << RunNum << "_" <<input_params.NumSubRun<< ".bin";
    mmsg.str("");
    mmsg<<"Checking for: "<<binarysubrunname.str();
    DANCE_Info("Main",mmsg.str());
    
    //Look for uncompressed .bin subrun files
    gz_in=gzopen(binarysubrunname.str().c_str(),"rb");
    
    //check to see if its open
    if(gz_in) {
      mmsg.str("");
      mmsg<<"File "<<binarysubrunname.str().c_str()<<" Found";
      DANCE_Success("Main",mmsg.str());
      runname << binarysubrunname.str();
      input_params.SubRunNumber=0;

      while(gz_in) {
        input_params.NumSubRun++;
        gz_queue.push(gz_in);
        binarysubrunname.str("");
        binarysubrunname << pathtodata << "/stage0_run_" << RunNum << "_" <<input_params.NumSubRun<< ".bin";
        gz_in=gzopen(binarysubrunname.str().c_str(),"rb");
      } 
    }
    else {
      //look for compressed .bin.gz subrun files
      binarysubrunname << ".gz";
      mmsg.str("");
      mmsg<<"Checking for: "<<binarysubrunname.str();
      DANCE_Info("Main",mmsg.str());
      gz_in=gzopen(binarysubrunname.str().c_str(),"rb");

      if(gz_in) {
	mmsg.str("");
	mmsg<<"File "<<binarysubrunname.str().c_str()<<" Found";
	DANCE_Success("Main",mmsg.str());
	runname << binarysubrunname.str();
        input_params.SubRunNumber=0;

        while(gz_in) {
          input_params.NumSubRun++;
          gz_queue.push(gz_in);
          binarysubrunname.str("");
          binarysubrunname << pathtodata << "/stage0_run_" << RunNum << "_" <<input_params.NumSubRun<< ".bin.gz";
          gz_in=gzopen(binarysubrunname.str().c_str(),"rb");
        } 

      }
      else {
        mmsg.str("");
        mmsg<<"Checking for: "<<binaryrunname.str().c_str();
        DANCE_Info("Main",mmsg.str());
        gz_in=gzopen(binaryrunname.str().c_str(),"rb");
        if (gz_in) {
          mmsg.str("");
          mmsg<<"File "<<binaryrunname.str().c_str()<<" Found";
          DANCE_Success("Main",mmsg.str());
          runname << binaryrunname.str();
          input_params.SubRunNumber=-1;
          gz_queue.push(gz_in);
        }
        else {
          binaryrunname << ".gz";
          mmsg.str("");
          mmsg<<"Checking for: "<<binaryrunname.str().c_str();
          gz_in=gzopen(binaryrunname.str().c_str(),"rb");
          if(gz_in) {
            DANCE_Info("Main",mmsg.str());
            DANCE_Success("Main",mmsg.str());
            runname << binaryrunname.str();
            input_params.SubRunNumber=-1;
            gz_queue.push(gz_in); 
          }
        }
      }
    }
  }

  //Simulated Data from GEANT4
  else if(input_params.Read_Binary == 0 && input_params.Read_Simulation == 1) {
    
    stringstream simulationrunname;
    simulationrunname.str();
    simulationrunname << STAGE0_SIM << "/"<< input_params.Simulation_File_Name <<".bin";

    mmsg.str("");
    mmsg<<"Checking for: "<<simulationrunname.str();
    DANCE_Info("Main",mmsg.str());
    
    //Look for uncompressed .bin files
    gz_in=gzopen(simulationrunname.str().c_str(),"rb");
    
    //check to see if its open
    if(gz_in) {
      mmsg.str("");
      mmsg<<"File "<<simulationrunname.str().c_str()<<" Found";
      DANCE_Success("Main",mmsg.str());
      runname << simulationrunname.str();
      gz_queue.push(gz_in);
      input_params.SubRunNumber=-1;
    }
    else {
      //look for compressed .bin.gz files
      simulationrunname << ".gz";
      mmsg.str("");
      mmsg<<"Checking for: "<<simulationrunname.str();
      DANCE_Info("Main",mmsg.str());
      gz_in=gzopen(simulationrunname.str().c_str(),"rb");
      if(gz_in) {

	mmsg.str("");
	mmsg<<"File "<<simulationrunname.str().c_str()<<" Found";
	DANCE_Success("Main",mmsg.str());
	runname << simulationrunname.str();
        gz_queue.push(gz_in);
        input_params.SubRunNumber=-1;
      }
    }
  }

  else {
    mmsg.str("");
    mmsg<<"Conflict Between Read_Binary (set to "<<input_params.Read_Binary<<") and Read_Simulation set to ("<<input_params.Read_Simulation<<"). Exiting";
    DANCE_Error("Main",mmsg.str());
    return -1;
  }
  
  if(gz_queue.empty()) {
    mmsg.str("");
    mmsg<<"File queue empty, this is bad.  Exiting.";
    DANCE_Error("Main",mmsg.str());
    return -1;
  }

  //Time profiling stuff
  struct timeval tv;  						// real time  
  double begin, end, time_elapsed;			// start,stop, elapsed time

  //start time
  gettimeofday(&tv,NULL); 
  begin=tv.tv_sec+(tv.tv_usec/1000000.0);

  //Initialize Unpacker
  func_ret = Initialize_Unpacker(input_params);
  if(func_ret) {
    return func_ret;
  }
  //initialize eventbuilder
  func_ret = Initialize_Eventbuilder(input_params);
  if(func_ret) {
    return func_ret;
  }
  //initialize calibrator
  func_ret = Initialize_Calibrator(input_params);
  if(func_ret) {
    return func_ret;
  }
  //initialize validator
  func_ret = Initialize_Validator(input_params);
  if(func_ret) {
    return func_ret;
  }
  //Initialize analyzer
  func_ret = Initialize_Analyzer(input_params);
  if(func_ret) {
    return func_ret;
  }

  //Launch the unpacker
  int events_analyzed=  Unpack_Data(gz_queue, begin, input_params, &analysis_params);

  mmsg.str("");
  mmsg<<"Analysis Complete. Analyzed: "<<events_analyzed<<" Entries";
  DANCE_Success("Main",mmsg.str());
  
  //Print the time elapsed for the entire process
  gettimeofday(&tv,NULL);  
  end=tv.tv_sec+(tv.tv_usec/1000000.0);
  time_elapsed = (double)(end-begin); ;

  mmsg.str("");
  mmsg<<"Time Elapsed: "<<time_elapsed<<" Seconds";
  DANCE_Info("Main",mmsg.str());

}
